gitpod-io/gitpod
# [server] fix leaking listeners

#### What it does

If the server was disposed before async listeners are connected then such listeners never captured and released by that leaking the server as well. This PR makes sure that:
- when the client connection is dropped then the server does not try to send message to it
- listeners are collected synchronously to avoid race condition with the server disposal

This PR is not enough though because there are similar issues in the downstream. 

#### How to test

- Start a workspace and then stop. If it works smoothly then listening to the instance update is still correct.
- Start a prebuild and check that the terminal has a proper output. It means that listening to headless logs works as well.
- After that make the memory snapshot as described below and inspect it for `GitpodServer`.
  - Make sure that you stopped all workspaces and closed all tabs for dev environments.
  - It assumes nobody else is using the dev environment in the meantime.
  - It is fine to see at least one instance of `GitpodServer`. It is usually a constructor.
- You can also take snapshots before running the workspace and after stopping the workspace. Such snapshots can be compared to find what kind of objects are allocated but not garbage collected.

#### How to make V8 memory snapshot

- kubectl exec in the server container
- find a process running the server
- make a memory snapshot: kill -s SIGUSR2 <pid>
- the server process will make the full gc and then write the dump to /tmp folder to a file with `.heapshapshot` extension
  - it can a while, you are better to watch server logs with kubectl logs -f
  - the specific filename can be found in logs as well
- cat this file into your workspace and then download
- open Chrome devtools, select memory tab, open the context menu in the right side bar (Profile section), select load and then choose the downloaded memory snapshot
